pkg/controllers/management/clusterprovisioner/helm_cleaner.go

```diff
@@ -0,0 +1,230 @@
+
+	"github.com/helm/helm-mapkubeapis/pkg/mapping"
+	"github.com/pkg/errors"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/mod/semver"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/release"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	apiMappings = &mapping.Metadata{
+		Mappings: []*mapping.Mapping{
+			{
+				DeprecatedAPI:    "apiVersion: policy/v1beta1\nkind: PodSecurityPolicy\n",
+				RemovedInVersion: "v1.25",
+				// TODO add all APIs we want to cover here. Perhaps store it in a ConfigMap so we don't need to change
+				// code for updates.
+			},
+		},
+	}
+)
+
+type ClientGetter struct {
+	k8sClient  kubernetes.Clientset
+	restConfig rest.Config
+}
+
+func (c ClientGetter) ToRESTConfig() (*rest.Config, error) {
+	return &c.restConfig, nil
+}
+
+func (c ClientGetter) ToDiscoveryClient() (discovery.CachedDiscoveryInterface, error) {
+	return memory.NewMemCacheClient(c.k8sClient.Discovery()), nil
+}
+
+func (c ClientGetter) ToRESTMapper() (meta.RESTMapper, error) {
+	return meta.NewDefaultRESTMapper(nil), nil
+}
+
+func (c ClientGetter) ToRawKubeConfigLoader() clientcmd.ClientConfig {
+	return &clientcmd.DefaultClientConfig
+}
+
+func NewClientGetter(k8sClient kubernetes.Interface, restConfig rest.Config) ClientGetter {
+	return ClientGetter{
+		k8sClient:  *k8sClient.(*kubernetes.Clientset),
+		restConfig: restConfig,
+	}
+}
+
+func (p *Provisioner) cleanupHelmReleases(cluster *v3.Cluster) error {
+	clusterManager := p.clusterManager
+	client, err := clusterManager.K8sClient(cluster.Name)
+	if err != nil {
+		return errors.Wrapf(err, "[cleanupHelmReleases] failed to obtain the Kubernetes client instance")
+	}
+
+	restConfig, err := clusterManager.RESTConfig(cluster.Name)
+	if err != nil {
+		return errors.Wrapf(err, "[cleanupHelmReleases] failed to obtain the Kubernetes REST config instance")
+	}
+
+	listNamespaces, err := client.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "[cleanupHelmReleases] failed to list namespaces on cluster %v", cluster.Name)
+	}
+
+	clientGetter := NewClientGetter(client, *restConfig)
+	helmDriver := os.Getenv("HELM_DRIVER")
+
+	for _, namespace := range listNamespaces.Items {
+		// TODO check if there's a way of doing this using a more API-based approach
+		actionConfig := new(action.Configuration)
+
+		if err = actionConfig.Init(
+			clientGetter,
+			namespace.Name,
+			helmDriver,
+			debugLog,
+		); err != nil {
+			return errors.Wrapf(err, "[cleanupHelmReleases] failed to create ActionConfiguration instance for Helm")
+		}
+
+		releases, err := actionConfig.Releases.ListReleases()
+		if err != nil {
+			return errors.Wrapf(err, "[cleanupHelmReleases] failed to list Helm releases for namespace %v", namespace.Name)
+		}
+
+		for _, helmRelease := range releases {
+			lastRelease, err := actionConfig.Releases.Last(helmRelease.Name)
```

Comment:
In addition, this also appears to only modify/replace the last release. I'm unsure how this will work to remove the rollback versions in the secret, although I expect I'm missing something there with the use of `superseded` below. If you could elaborate more on the mechanism (either here on in a comment) I would appreciate it.

Reply:
> It does upgrade a Helm release, yes, but as far as I know there is no other way around it.

Per our conversation in another thread, that makes sense to me now, thanks.

> It does not modify the contents of the latest release version. Rather, it creates a new release version with the modifications required. I am not an expert on Helm but I do believe it does not prevent a rollback, but a rollback would likely not work anyways because there are lingering PSP resources (that cannot be applied anymore).

Ok, thanks for explaining. So this is exclusively for situations where you upgraded to k8s 1.25, but the helm release still contained an active PSP right? 

So something like:
- I have a k8s 1.24 cluster
- I install a chart, it uses a PSP to run
- I upgrade to k8s 1.25
- Now I can't upgrade the chart, since it can't render the currently installed version (no psp's)

> As for PSAs breaking an app, I believe the exemptions would address this. I could additionally not upgrade Helm applications that are not in namespaces Rancher controls. What do you think?

A couple of things here:
- The exemptions will only apply in downstream clusters (AFAIK). We also need to consider the local cluster.
- As you noted, this will also apply to user namespaces, which will have no such exception. 

Limiting this to our resources is probably a smart move from a safety standpoint, especially if we end up including more than PSPs in our list of resources-to-remove. I like that suggestion - maybe we can import the same list that the exceptions logic uses.

>  This process does not run unless the cluster is in Kubernetes v1.25, so I wouldn't think it's dropping PSPs in a cluster that still supports it:

My concern is more that we are removing a buffer period where users should be making their apps compatible with 1.25. I would expect the upgrade process to go like this:

1. A user is on 1.24
2. They see that 1.25 removes PSP support, but keeps PSA support which can use in 1.24
3. They examine their installed apps (charts and non-charts), and upgrade them from PSP to PSA/something else, and validate that the functionality works
4. They upgrade to 1.25

This approach removes step 3, and instead encourages a "delete and figure it out later" approach. This way, the user can just do a `k delete psp` for everything, and upgrade. But that won't work in every case, since they didn't upgrade their app to work, they just removed the things that were preventing the upgrade. If the user had followed the approach outlined above I don't think we would need this functionality (correct me if I am wrong there, I may still be missing something). 

Though this is all based on my understanding - I could be wrong there so definitely open to discussion.

Reply:
I'm not entirely following. I don't see how this code here prevents a user from running your upgrade scenario.
This code here is meant to address the scenario where a chart was present and is now stuck because of lingering PSPs. In this specific case, your chart gets stuck -- you cannot upgrade and/or delete, because Helm keeps trying to reapply a resource for which the API server doesn't have a record anymore.

Reply:
The code allows the user to upgrade to 1.25 without running the upgrade scenario. This can cause the app to break since a PSP (critical to the app running) was removed and the user never upgraded their app to handle it.

If the code didn't exist, the user would be forced to migrate their application to a newer version before upgrading.

It's more that the use case that this is serving is cases where you didn't properly upgrade. I think we can justify it for our things (Although we still need to consider it for the local cluster) since we control the upgrade process for our stuff, but it's harder to make that case for our users.

Reply:
re: Customers not following a process of seeing a failure to know that they should migrate from PSPs to PSAs; absolutely agree @MbolotSuse, good catch. 

I agree with @stormqueen1990 here on her comments overall, but would point out that IMO this controller should be **enabled by Feature Flag, disabled by default**; our more security-conscious users should choose to migrate manually on seeing failures on upgrading to 1.25, v.s. others who do not operate hardened setups (where PSPs are still deployed, but are effectively unused) should use this controller. I believe currently it is always on?

Also re:

> The code allows the user to upgrade to 1.25 without running the upgrade scenario. This can cause the app to break since a PSP (critical to the app running) was removed and the user never upgraded their app to handle it.
> 
> If the code didn't exist, the user would be forced to migrate their application to a newer version before upgrading (or at minimum would be responsible for forcibly bypassing any errors).

On upgrading from 1.24 -> 1.25, the definition of the PSP resource would anyways be removed from the API, so if an app is broken without PSPs existing, it would be broken before this code ever ran.

Regardless of whether this code existed or not, the user would be forced to migrate their application to a newer version, but without this code they wouldn't even be able to migrate (since running a helm upgrade involves deleting resources that aren't in the next helm release, which is where you would fail on trying to delete a non-existent API i.e. PSPs).